File: 문자열/leetcode205(Easy)-Isomorphic Strings.js
// 79ms(48%), 45MB(20%)

const indexing = (string) => {
  const table = {}
  Array.from(string).forEach((char, index) => {
    if (!table[char]) {
      table[char] = index
    }
  })
  return Array.from(string).map((char) => table[char])
}

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function (s, t) {
  const n = s.length
  const ss = indexing(s)
  const tt = indexing(t)

  for (let i = 0; i < n; i += 1) {
    if (ss[i] !== tt[i]) {
      return false
    }
  }
  return true
}

File: 문자열/leetcode205(Easy)-Isomorphic Strings.py
# 46ms(50%), 14MB(12%)
class Solution:
    def indexing(self, string):
        table = {}
        for index, char in enumerate(string):
            if char not in table:
                table[char] = index
        return [table[char] for char in string]

    def isIsomorphic(self, s: str, t: str) -> bool:
        n = len(s)
        ss = self.indexing(s)
        tt = self.indexing(t)
        for i in range(n):
            if ss[i] != tt[i]:
                return False

        return True